x/airdrop/keeper/claim_test.go

```diff
@@ -0,0 +1,450 @@
+			s.SetupTest() // reset state
+
+			claimer := s.TestAccs[0]
+			distributor := s.TestAccs[1]
+
+			// Fund the distributor
+			initialDistributorBalance := sdk.NewInt(1000)
+			s.FundAccount(distributor, sdk.NewCoin(RewardDenom, initialDistributorBalance))
+
+			// Create the initial airdrop config
+			s.App.AirdropKeeper.SetAirdrop(s.Ctx, types.Airdrop{
+				Id:                    AirdropId,
+				RewardDenom:           RewardDenom,
+				DistributionAddress:   distributor.String(),
+				DistributionStartDate: &DistributionStartDate,
+				ClaimTypeDeadlineDate: &DeadlineDate,
+				EarlyClaimPenalty:     tc.penalty,
+			})
+
+			// Set the block time to the distribution start time plus the offset
+			blockTime := DistributionStartDate.Add(tc.timeOffset)
+			s.Ctx = s.Ctx.WithBlockTime(blockTime)
+
+			// Create the initial user and allocations
+			s.App.AirdropKeeper.SetUserAllocation(s.Ctx, types.UserAllocation{
+				AirdropId:   AirdropId,
+				Address:     claimer.String(),
+				ClaimType:   tc.initialClaimType,
+				Claimed:     sdkmath.NewInt(tc.initialClaimed),
+				Forfeited:   sdkmath.ZeroInt(),
+				Allocations: allocationsToSdkInt(tc.initialAllocations),
+			})
+
+			// Call claim daily
+			actualError := s.App.AirdropKeeper.ClaimEarly(s.Ctx, AirdropId, claimer.String())
+			if tc.expectedError != "" {
+				s.Require().ErrorContains(actualError, tc.expectedError)
+				return
+			}
+			s.Require().NoError(actualError, "no error expected when claiming early")
+
+			// Check that the user was updated
+			userAllocation := s.MustGetUserAllocation(AirdropId, claimer.String())
+			s.Require().Equal(tc.expectedClaimed, userAllocation.Claimed.Int64(), "claimed")
+			s.Require().Equal(tc.expectedForfeited, userAllocation.Forfeited.Int64(), "forfeited")
+			s.Require().Equal(types.CLAIM_EARLY, userAllocation.ClaimType, "claim types")
+			for _, allocation := range userAllocation.Allocations {
+				s.Require().Zero(allocation.Int64(), "allocations should be 0")
+			}
+
+			// Confirm funds were decremented from the distributor
+			expectedDistributorBalance := initialDistributorBalance.Sub(sdkmath.NewInt(tc.expectedUserBalanceChange))
+			actualDistributorBalance := s.App.BankKeeper.GetBalance(s.Ctx, distributor, RewardDenom).Amount
+			s.Require().Equal(expectedDistributorBalance.Int64(), actualDistributorBalance.Int64(), "distributor balance")
+
+			// Confirm funds were sent to the user
+			claimerBalance := s.App.BankKeeper.GetBalance(s.Ctx, claimer, RewardDenom).Amount
+			s.Require().Equal(tc.expectedUserBalanceChange, claimerBalance.Int64(), "claimer balance")
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestLinkAddresses() {
+	testCases := []struct {
+		name                string
+		initialClaimType    types.ClaimType
+		initialClaimed      sdkmath.Int
+		strideAllocations   []int64
+		hostAllocations     []int64
+		expectedAllocations []int64
+		expectedError       string
+	}{
+		{
+			name:                "no stride allocations",
```

Comment:
Is there a reverse case where StrideAllocations is non-nil and hostAllocations is nil or all zeros?

Reply:
just added a test case for nil host! (also covers the case where both are nil since it errors if host is nil)